pkg/vault/contracts/WrappedBalancerPoolToken.sol

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { ERC20Permit } from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Permit.sol";
+
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+import { IWrappedBalancerPoolToken } from "@balancer-labs/v3-interfaces/contracts/vault/IWrappedBalancerPoolToken.sol";
+
+/**
+ * @notice ERC20 wrapper for Balancer Pool Token (BPT), allowing users to deposit BPT
+ * and receive 1:1 wrapped tokens, or burn wrapped tokens to redeem the original BPT.
+ * Minting and burning are only allowed when the Vault is locked.
+ */
+contract WrappedBalancerPoolToken is IWrappedBalancerPoolToken, ERC20, ERC20Permit {
```

Comment:
The most important invariant of this contract is that every time we mint / burn, `bpt.balanceOf(this) == totalSupply`. 

Do you think it's worth adding that (redundant) check after each operation? E.g.

```
modifier fullyBacked() {
    _;
    if (bpt.balanceOf(address(this)) < totalSupply()) {
        revert NotFullyBacked();
    }
}
```

Reply:
note the use of `<` instead of `!=` as anyone can donate BPT to the wrapper contract. The point is that locked assets should be _at least_ 1:1 with the total supply; if we have more assets than total supply that's also fine.

Reply:
I think it is worth it for piece of mind, given that this shouldn't be a "critical path" function, so we don't care too much about gas (and the check is pretty cheap anyway). We'd just have to make sure legitimate deposits can't get locked somehow.

Reply:
I think this check is pointless, because if we can prove through tests that 1 wBPT = 1 BPT, then there won't be any cases where it's possible to break it.

Since BPT must be owned by the Vault, the only way to create malicious BPTs would be to make a pool where BPTs can be freely minted, but cannot be burned if they belong to another participant. That means WBPT would never actually reach the scenario where this fullyBacked condition fails.